src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KDF.java

```diff
@@ -0,0 +1,340 @@
+            default -> throw new InvalidAlgorithmParameterException(
+                    "derivationSpec must be a " + HKDFParameterSpec.class +
+                    " instance, instead of " + derivationSpec.getClass());
+        }
+
+        P11Key p11BaseKey = convertKey(baseKey, (isExtract ? "IKM" : "PRK") +
+                " could not be converted to a token key for HKDF derivation.");
+
+        long saltType = CKF_HKDF_SALT_NULL;
+        byte[] saltBytes = null;
+        P11Key p11SaltKey = null;
+        if (salt instanceof SecretKeySpec) {
+            saltType = CKF_HKDF_SALT_DATA;
+            saltBytes = salt.getEncoded();
+        } else if (salt != null) {
+            // consolidateKeyMaterial returns a salt from the token.
+            saltType = CKF_HKDF_SALT_KEY;
+            p11SaltKey = (P11Key.P11SecretKey) salt;
+            assert p11SaltKey.token == token : "salt must be from the same " +
+                    "token as service.";
+        }
+
+        P11SecretKeyFactory.KeyInfo ki = P11SecretKeyFactory.getKeyInfo(alg);
+        if (ki == null) {
+            throw new InvalidAlgorithmParameterException("A PKCS #11 key " +
+                    "type (CKK_*) was not found for a key of the algorithm '" +
+                    alg + "'.");
+        }
+        long derivedKeyType = ki.keyType;
+        long derivedKeyClass = isData ? CKO_DATA : CKO_SECRET_KEY;
+        CK_ATTRIBUTE[] attrs = new CK_ATTRIBUTE[] {
+                new CK_ATTRIBUTE(CKA_CLASS, derivedKeyClass),
+                new CK_ATTRIBUTE(CKA_KEY_TYPE, derivedKeyType),
+                new CK_ATTRIBUTE(CKA_VALUE_LEN, outLen)
+        };
+        Session session = null;
+        long baseKeyID = p11BaseKey.getKeyID();
+        try {
+            session = token.getOpSession();
+            CK_HKDF_PARAMS params = new CK_HKDF_PARAMS(isExtract, isExpand,
+                    hmacMechanism, saltType, saltBytes, p11SaltKey != null ?
+                    p11SaltKey.getKeyID() : 0L, info);
+            attrs = token.getAttributes(O_GENERATE, derivedKeyClass,
+                    derivedKeyType, attrs);
+            long derivedObjectID = token.p11.C_DeriveKey(session.id(),
+                    new CK_MECHANISM(isData ? CKM_HKDF_DATA : CKM_HKDF_DERIVE,
+                    params), baseKeyID, attrs);
+            Object ret;
+            if (isData) {
+                try {
+                    CK_ATTRIBUTE[] dataAttr = new CK_ATTRIBUTE[] {
+                            new CK_ATTRIBUTE(CKA_VALUE)
+                    };
+                    token.p11.C_GetAttributeValue(session.id(), derivedObjectID,
+                            dataAttr);
+                    ret = dataAttr[0].getByteArray();
+                } finally {
+                    token.p11.C_DestroyObject(session.id(), derivedObjectID);
+                }
+            } else {
+                ret = P11Key.secretKey(session, derivedObjectID, alg, outLen,
```

Comment:
Is the key length in bytes or bits here?

Reply:
Good question. Should be converted to bits. The same issue repeats one more time.